### ISSUE TITLE
Sharing automounted USB with 'include'

### ISSUE THREAD

**Issue by fawzibr** [ServerContainers/samba#144]
I was thinking of using samba as a container but don't know how to do this part or if there is a way to do it.

In my current samba install (smb.conf) I have the following:

```
include = /etc/samba/usb-shares.conf
```

After some UDEV magic when the USB is connected I add the mount share to /etc/samba/usb-shares.conf and call 'smbcontrol smbd reload-config' and it shares the USB. When the USB is removed, it removes the share and reloads the config again.

I was thinking of doing the following:

```
services:
  samba:
    container_name: samba
    image: ghcr.io/servercontainers/samba
    restart: always
    network_mode: host
    cap_add:
      - CAP_NET_ADMIN
    environment:
      SAMBA_CONF_WORKGROUP: WORKGROUP
      SAMBA_CONF_SERVER_STRING: Samba Server (Docker)
      SAMBA_CONF_SERVER_ROLE: standalone server
      SAMBA_CONF_MAP_TO_GUEST: Bad User
      SAMBA_CONF_LOG_LEVEL: 3
      SAMBA_GLOBAL_CONFIG_wins_SPACE_support: yes
      SAMBA_GLOBAL_CONFIG_follow_SPACE_symlinks: yes
      SAMBA_VOLUME_CONFIG_storage: "[storage]; path=/shares/storage/; browseable = yes; writeable = yes; create mask = 0777; directory mask = 0777; guest ok = yes; guest only = yes"

      # this adds the include file to the smb.conf 
      SAMBA_VOLUME_CONFIG_includes: "include /shares/includes/usb-mounts.conf"
    volumes:
      - /etc/avahi/services/:/external/avahi
      - /storage:/shares/storage
      # this volume will contains the include files
      - /includes-dir:/shares/includes
      # this volume is where UDEV would mount the USBs
      - /usb-mount-dir:/shares/usb
```     

I imagine that would work, my only problem would be the reload config part. So is there a way to reload the config, or have a way to reload the config when the include files changes?

**Comment by MarvAmBass** [comment]
interesting question. multiple solutions:

simplest (might not work): use `docker exec <runningContainerName/ID> smbcontrol smbd reload-config`
this might work, so your `usb-mounts.conf` gets updated on the host, and you trigger the reload inside the container.

another hard method would be to `docker-compose down; docker-compose up` but this will lead to a downtime.

you might also be able to put the udev magic into the container - but I'd advise to create your own container then...

**Comment by MarvAmBass** [comment]
I'll close this since it's more of a question which I answered - but there are many other solutions - I also don't think that you have much benefit of using my container - it's a very specific setup you might want to create a smb container which does exactly that from scratch.

**Comment by fawzibr** [comment]
Well I worked on it and instead of using includes (the reload-config did't work) I tried usershares and that worked. Some simple small changes were made:

- scripts/entrypoint.sh - Add usershare stuff to smb.conf, and set a crontab to check the usershare directory. I commented my change in entrypoint.sh so is easy to find.
- scripts/check-usershare.sh - checks directory for *.usershare files, if anything changes it adds/deletes the usershares
- config/runit/crond/run - start crond 

Added new environment variable SAMBA_USERSHARES_DIR, every file named "<share_name>.usershare" will be mounted as a new share. It checks for changes every minute.

Usershare file format (check samba documentation for the values - net usershare):
path=
comment= <- optional
acl= <- optional
guest_ok= <- optional

I'm attaching a file with the changes if you are interested and want to test it. It doesn't really change much. 

[samba-master-changes.zip](https://github.com/user-attachments/files/17331942/samba-master-changes.zip)
